src/ORM/Rule/LinkConstraint.php

```diff
@@ -0,0 +1,260 @@
+        }
+        return $fields;
+    }
+
+    /**
+     * Build conditions.
+     *
+     * @param array $fields The condition fields.
+     * @param array $values The condition values.
+     * @return array A conditions array combined from the passed fields and values.
+     */
+    protected function _buildConditions($fields, $values)
+    {
+        if (count($fields) !== count($values)) {
+            throw new \InvalidArgumentException('The number of fields is expected to match the number of values.');
+        }
+        return array_combine($fields, $values);
+    }
+
+    /**
+     * Count links.
+     *
+     * @param \Cake\ORM\Association $association The association for which to count links.
+     * @param \Cake\Datasource\EntityInterface $entity The entity involved in the operation.
+     * @return int|null The number of links, or `null` for unsupported association types.
+     */
+    protected function _countLinks($association, $entity)
+    {
+        $count = null;
+
+        switch ($association->type()) {
+            case Association::ONE_TO_ONE:
+            case Association::ONE_TO_MANY:
+                $conditions = $this->_buildConditions(
+                    $this->_aliasFields((array)$association->foreignKey(), $association->target()),
+                    $entity->extract((array)$association->bindingKey())
+                );
+
+                $count = $association
+                    ->target()
+                    ->find()
+                    ->where($conditions)
+                    ->count();
+                break;
+
+            case Association::MANY_TO_MANY:
+                $primaryKey = $association->source()->primaryKey();
+                $conditions = $this->_buildConditions(
+                    $this->_aliasFields((array)$primaryKey, $association->source()),
+                    $entity->extract((array)$primaryKey)
+                );
+
+                $sourceAlias = $association->source()->registryAlias();
+                $sourceAssociation = $association->target()->association($sourceAlias);
+
+                $count = $association
+                    ->target()
+                    ->find()
+                    ->matching(
+                        $sourceAssociation->registryAlias(),
+                        function (Query $query) use ($conditions) {
+                            return $query->where($conditions);
+                        }
+                    )
+                    ->count();
+                break;
+
+            case Association::MANY_TO_ONE:
+                $conditions = $this->_buildConditions(
+                    $this->_aliasFields((array)$association->bindingKey(), $association->target()),
+                    $entity->extract((array)$association->foreignKey())
+                );
+
+                $count = $association
+                    ->target()
+                    ->find()
+                    ->where($conditions)
```

Comment:
Now I remember why I was doing things the way I was doing them, I should have added tests for this.

With custom finders/conditions involved that make references to the main/parent table, for example in a `Articles hasMany Comments` situation, where the `Comments` association conditions refer to a `Articles` column, like `Comments.foo = Articles.bar`, things will break when counting on `Comments` as `Articles` isn't included.

Note sure if this is what you where up to @markstory , but theoretically it should be possible to use the "query from the other side using matching" approch for all association types (given there are proper associations defined for both sides of course), shouldn't it? Despite being possibly rather unefficient, I may be overseeing something essential, it's very late over here and this sounds a little too easy.

Another somewhat simple, but possibly also more unefficient approach, would be to do a find on the main repository with the configured association contained, and then simply check the results.